src/test/util/setup_common.cpp

```diff
@@ -124,76 +124,88 @@ BasicTestingSetup::~BasicTestingSetup()
     }
     g_parallel_script_checks = true;
 
-    m_node.banman = MakeUnique<BanMan>(GetDataDir() / "banlist.dat", nullptr, DEFAULT_MISBEHAVING_BANTIME);
-    m_node.connman = MakeUnique<CConnman>(0x1337, 0x1337); // Deterministic randomness for tests.
-    m_node.peerman = MakeUnique<PeerManager>(chainparams, *m_node.connman, m_node.banman.get(), *m_node.scheduler, *m_node.chainman, *m_node.mempool);
-    {
-        CConnman::Options options;
-        options.m_msgproc = m_node.peerman.get();
-        m_node.connman->Init(options);
+    if (!chain_testing_only) {
+        const CChainParams& chainparams = Params();
+        // Ideally we'd move all the RPC tests to the functional testing framework
+        // instead of unit tests, but for now we need these here.
+        RegisterAllCoreRPCCommands(tableRPC);
+
+        node.chainman->InitializeChainstate(*node.mempool);
+        ::ChainstateActive().InitCoinsDB(
+                                         /* cache_size_bytes */ 1 << 23, /* in_memory */ true, /* should_wipe */ false);
+        assert(!::ChainstateActive().CanFlushToDisk());
+        ::ChainstateActive().InitCoinsCache(1 << 23);
+        assert(::ChainstateActive().CanFlushToDisk());
+        if (!LoadGenesisBlock(chainparams)) {
+            throw std::runtime_error("LoadGenesisBlock failed.");
+        }
+
+        BlockValidationState state;
+        if (!ActivateBestChain(state, chainparams)) {
+            throw std::runtime_error(strprintf("ActivateBestChain failed. (%s)", state.ToString()));
+        }
+
+        node.banman = MakeUnique<BanMan>(GetDataDir() / "banlist.dat", nullptr, DEFAULT_MISBEHAVING_BANTIME);
+        node.connman = MakeUnique<CConnman>(0x1337, 0x1337); // Deterministic randomness for tests.
+        node.peerman = MakeUnique<PeerManager>(chainparams, *node.connman, node.banman.get(), *node.scheduler, *node.chainman, *node.mempool);
+        {
+            CConnman::Options options;
+            options.m_msgproc = node.peerman.get();
+            node.connman->Init(options);
+        }
     }
 }
 
-TestingSetup::~TestingSetup()
-{
-    if (m_node.scheduler) m_node.scheduler->stop();
-    threadGroup.interrupt_all();
-    threadGroup.join_all();
+void TestingAppDestructionSequence(NodeContext& node, boost::thread_group& thread_group, bool chain_testing_only) {
+    if (node.scheduler) node.scheduler->stop();
+    thread_group.interrupt_all();
+    thread_group.join_all();
     GetMainSignals().FlushBackgroundCallbacks();
     GetMainSignals().UnregisterBackgroundSignalScheduler();
-    m_node.connman.reset();
-    m_node.banman.reset();
-    m_node.args = nullptr;
-    UnloadBlockIndex(m_node.mempool.get(), *m_node.chainman);
-    m_node.mempool.reset();
-    m_node.scheduler.reset();
-    m_node.chainman->Reset();
-    m_node.chainman = nullptr;
+    if (!chain_testing_only) {
+        node.connman.reset();
+        node.banman.reset();
+    }
```

Comment:
any reason for the modifications here? reset is idempotent , so doesn't need to be guarded.